teuthology/test/task/__init__.py

```diff
@@ -0,0 +1,132 @@
+                task_hosts = task.hosts
+                assert len(task_hosts) == 0
+
+    def test_hosts_one_role(self):
+        self.ctx.cluster = Cluster()
+        self.ctx.cluster.add(Remote('remote1'), ['role1'])
+        self.ctx.cluster.add(Remote('remote2'), ['role2'])
+        task_config = dict(
+            hosts=['role1'],
+        )
+        with Task(self.ctx, task_config) as task:
+            task_hosts = task.hosts
+            assert len(task_hosts) == 1
+            assert task_hosts[0].name == 'remote1'
+
+    def test_hosts_two_roles(self):
+        self.ctx.cluster = Cluster()
+        self.ctx.cluster.add(Remote('remote1'), ['role1'])
+        self.ctx.cluster.add(Remote('remote2'), ['role2'])
+        self.ctx.cluster.add(Remote('remote3'), ['role3'])
+        task_config = dict(
+            hosts=['role1', 'role3'],
+        )
+        with Task(self.ctx, task_config) as task:
+            task_hosts = task.hosts
+            assert len(task_hosts) == 2
+            hostnames = [host.name for host in task_hosts]
+            assert sorted(hostnames) == ['remote1', 'remote3']
+
+    def test_hosts_two_hostnames(self):
+        self.ctx.cluster = Cluster()
+        self.ctx.cluster.add(Remote('remote1.example.com'), ['role1'])
+        self.ctx.cluster.add(Remote('remote2.example.com'), ['role2'])
+        self.ctx.cluster.add(Remote('remote3.example.com'), ['role3'])
+        task_config = dict(
+            hosts=['remote1', 'remote2.example.com'],
+        )
+        with Task(self.ctx, task_config) as task:
+            task_hosts = task.hosts
+            assert len(task_hosts) == 2
+            hostnames = [host.name for host in task_hosts]
+            assert sorted(hostnames) == ['remote1.example.com',
+                                         'remote2.example.com']
+
+    def test_hosts_one_role_one_hostname(self):
+        self.ctx.cluster = Cluster()
+        self.ctx.cluster.add(Remote('remote1.example.com'), ['role1'])
+        self.ctx.cluster.add(Remote('remote2.example.com'), ['role2'])
+        self.ctx.cluster.add(Remote('remote3.example.com'), ['role3'])
+        task_config = dict(
+            hosts=['role1', 'remote2.example.com'],
+        )
+        with Task(self.ctx, task_config) as task:
+            task_hosts = task.hosts
+            assert len(task_hosts) == 2
+            hostnames = [host.name for host in task_hosts]
+            assert sorted(hostnames) == ['remote1.example.com',
+                                         'remote2.example.com']
+
+    @patch.object(Task, 'setup')
+    def test_setup_called(self, m_setup):
+        task_config = dict()
+        with Task(self.ctx, task_config):
+            m_setup.assert_called_once_with()
+
+    @patch.object(Task, 'begin')
+    def test_begin_called(self, m_begin):
+        task_config = dict()
+        with Task(self.ctx, task_config):
+            m_begin.assert_called_once_with()
```

Comment:
I want the asserts to be evaluated as soon as possible; this one can be done inside the ``with`` block but the other one cannot.